### ISSUE TITLE
Support for 'use-sharp-yuv' with WebP?

### ISSUE THREAD

**Issue by jethrodesign** [Imagick/imagick#390]
Hi, I'm just curious to know if Imagick supports passing the 'use-sharp-yuv' option for WebP format if the underlying install of ImageMagick supports it (see: https://imagemagick.org/script/webp.php)? Can't find any resources or information about this anywhere.

Our server is running CloudLinux 7.9 (and cPanel) and this gives us Imagick 3.4.4 with ImageMagick 7.0.10-10 Q16, which is compiled with WebP support.

Unfortunately, with the way our server is configured and with how the plugins in the sites we're supporting work, calling ImageMagick directly through command line is not really an option. We need to use Imagick.

Thanks for any insight here.

**Comment by Danack** [comment]
Most of the options can be set through the the set option method, but they also require knowing the appropriate config option which....isn't always obvious from the docs.

Looking at the code, try:

```
$imagick->setOption("webp:use-sharp-yuv", "true");
```

And let me know if that works?

**Comment by WideGal** [comment]
I have the same exact question as the OP. Tried Danack's idea of both "true" and "1"; neither seem to do anything.

By contrast, generating the same image with use-sharp-yuv in the WebP Converter app does have a noticeable positive effect, and changes the file size slightly. I'd love to get this working on my server.

Any ideas?

**Comment by Danack** [comment]
> Any ideas?

Please can you (or someone) post:
i) An appropriate source image.
ii) The exact command line invocation used.
iii) The example output image that looks better.

I can trace it through the code, but I have no idea what this feature is meant to look like, so can't tell if it's working or not.

**Comment by WideGal** [comment]
Hi Danack,

Here is a [source jpg image](https://www.widerangegalleries.org/webp/Source.jpg). 326.4kb

When I output via the "WebP Converter" app on my computer (just for reference), with: Quality: 85, lossy, Method: 6, Sharp YUV: enabled, here is [the webp result](https://www.widerangegalleries.org/webp/WebP_Converter.webp). 178.3kb

And when I output via iMagick (php) on my server, here is [the iMagick webp result](https://www.widerangegalleries.org/webp/iMagick.webp). 164kb

When comparing these two webp images, particularly in the upper right cloud area you can see some subtle banding/boxing artifacts in the iMagick version which have been smoothed over in the WebP Converter version. I know it's subtle, but it's noticeable for pixelpeepers.

I believe this difference is because iMagick isn't successfully applying the use-sharp-yuv option. When I generate the iMagick webp files, the resulting image and file size  is exactly the same regardless of the use-sharp-yuv setting, whether it's on or off.

Here is my relevant php code for iMagick:

`$img = new Imagick($targetFile);
$img->setImageFormat('webp');
$img->setOption('webp:thread-level', '1');
$img->setOption('webp:low-memory', '0');
$img->setOption('webp:lossless', 'false');
$img->setOption('webp:method', '6');
$img->setOption('webp:alpha-filtering', '0');
$img->setOption('webp:auto-filter', 'true');
$img->setOption('webp:use-sharp-yuv', 'true');
$img->setImageCompressionQuality(85);
$img->writeImage( $filepath . $filename .'.webp' );`

Thanks for taking a look!

**Comment by jethrodesign** [comment]
Hi. Still following the developments here and trying to do testing on my end.

If it's helpful, [here's an article ](https://titanwolf.org/Network/Articles/Article?AID=65aba585-2bd1-4977-8798-7db7502a7b1c#gsc.tab=0)that describes the specific issue that WebPs have and how 'sharp-yuv' helps overcome it.  Note that it is primarily an issue that affects thin lines and vibrant colors, though I'm sure it may have a bit of an affect in other areas as WideGal is noticing. 

I have a sample image that shows the issue VERY clearly and obviously. I'm just figuring out where/how to post it. And I also need to install the libwebp & cwebp libraries locally so I can actually run a test using 'sharp-yuv'. I currently only have the WebP conversion that does NOT use 'sharp-yuv', and looks like garbage...

**Comment by Danack** [comment]
> When I output via the "WebP Converter" 

@WideGal Although that might give different, and superior, output than ImageMagick/Imagick, this issue under investigation here is just making sure that flag is passed through Imagick correctly. To do that comparing it to images passed through the ImageMagick convert tool, like: 


```
% convert rb-grid.png rb-grid.webp

% convert rb-grid.png -define webp:use-sharp-yuv=1 rb-grid-sharp-yuv.webp

% convert rb-grid.png -sampling-factor 4: 2: 0 rb-grid-YUV420.jpg # (Reference)
```

compared to images passed through Imagick, are the ones I'll be looking at.

Inconveniently, it seems that compiling ImageMagick from src appears to have a bug in it's installer:

```
# convert -v
convert: error while loading shared libraries: libMagickCore-7.Q16HDRI.so.6: cannot open shared object file: No such file or directory
```
which I'll need to investigate first...

**Comment by Danack** [comment]
tl:dr It appears to be enabled by default. At least for the webp library that I'm using.

The code below generates some test images, and they all seem to be identical (or at least perceptually the same), to the one generated by `convert rb-grid.png -define webp:use-sharp-yuv=1 rb-grid-sharp-yuv.webp`

As github doesn't support webp images, showing this is kind of annoyingly hard. I tried to make a gif or png with the outputs, but it seems either method results in the differences being erased.

So here is a screenshot of the images `rb-grid-imagick-default-sharp.webp` `rb-grid-imagick-use-sharp-yuv_true.webp` and `rb-grid-imagick-use-sharp-yuv_false.webp`

![Screen Shot 2021-05-19 at 21 09 10](https://user-images.githubusercontent.com/1505719/118877805-c7435680-b8e6-11eb-827b-6838a2bb6ff5.png)

It's easier to see when switching between them fast, but the rb-grid-imagick-use-sharp-yuv_false.webp one that used  `$img->setOption('webp:use-sharp-yuv', "false");` has distinctly less vibrant red lines.

I'm think I'm going to close this issue.

If anyone still thinks that they are lacking  the use-sharp-yuv setting, even after they have set it through something like `$img->setOption('webp:use-sharp-yuv', "true");`, please open a new issue, with a convert command line equivalent.


The test code was:

```
<?php

declare(strict_types = 1);


$targetFile = __DIR__ . "/Source.jpg";

$img = new Imagick("rb-grid.png");
$img->setImageFormat('webp');

$img->writeImage(__DIR__ . '/rb-grid-imagick-default-sharp.webp');

$img->setOption('webp:use-sharp-yuv', "1");
$img->writeImage(__DIR__ . '/rb-grid-imagick-use-sharp-yuv_1.webp');

$img->setOption('webp:use-sharp-yuv', "0");
$img->writeImage(__DIR__ . '/rb-grid-imagick-use-sharp-yuv_0.webp');

$img->setOption('webp:use-sharp-yuv', "true");
$img->writeImage(__DIR__ . '/rb-grid-imagick-use-sharp-yuv_true.webp');

$img->setOption('webp:use-sharp-yuv', "false");
$img->writeImage(__DIR__ . '/rb-grid-imagick-use-sharp-yuv_false.webp');


// This is completely useless. The colour differences
// are invisible after writing the png
//
//$files = [
//    "0" => __DIR__ . '/rb-grid-imagick-use-sharp-yuv_0.webp',
//    "1" => __DIR__ . '/rb-grid-imagick-use-sharp-yuv_1.webp',
//    "false" => __DIR__ . '/rb-grid-imagick-use-sharp-yuv_false.webp',
//    "true" => __DIR__ . '/rb-grid-imagick-use-sharp-yuv_true.webp',
//];

//$draw = new \ImagickDraw();
//$draw->setStrokeColor("black");
//$draw->setFillColor("white");
//
//$draw->setStrokeWidth(1);
//$draw->setFontSize(16);
//
//$canvas = new Imagick();
//$canvas->newPseudoImage(256 * 2, 256 * 2, "xc:black");
//$canvas->setFormat("png");
//
//$count = 0;
//foreach ($files as $name => $file) {
//
//    echo "Adding file $file \n";
//
//    $text = "use-sharp-yuv = " . $name;
//    $image = new Imagick($file);
//    $image->annotateimage($draw, 40, 40, 0, $text);
//    $canvas->compositeImage(
//        $image, \Imagick::COMPOSITE_ATOP,
//        ($count % 2) * 256,
//        (($count /2) % 2) * 256,
//    );
//
//    $image->destroy();
//    $count += 1;
//}
//
//
//$canvas->writeImage("png24:output.png");
//


echo "fin\n";

```

**Comment by Danack** [comment]
To note, it is slightly mysterious that it is enabled by default. It's probably being done through the function `WebPConfigInit`, which doesn't appear to set it in the default spec, but I can't easily check what's happening on the actual version I'm running against. 

but `$img->setOption('webp:use-sharp-yuv', "false");` definitely makes the image quality worse.

Also, Google are not great at keeping pages live: http://webp.googlecode.com/
